cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java

```diff
@@ -215,65 +264,105 @@ public Void call(DatasetContext context) throws Exception {
+        @Override
+        public List<Message> call(DatasetContext context) throws Exception {
+          return doFetchMessages(messageId);
+        }
+      }, TopicNotFoundException.class, IOException.class);
     }
 
-    AppMetadataStore getAppMetadataStore(DatasetContext context) {
-      // TODO Find a way to access the appMetadataStore without copying code from the DefaultStore
-      try {
-        Table table = DatasetsUtil.getOrCreateDataset(context, datasetFramework, APP_META_INSTANCE_ID,
-                                                      Table.class.getName(), DatasetProperties.EMPTY);
-        AppMetadataStore appMetadataStore = new AppMetadataStore(table, cConf, upgradeComplete);
-        // If upgrade was not complete, check if it is and update boolean
-        if (!upgradeComplete.get()) {
-          boolean isUpgradeComplete = appMetadataStore.isUpgradeComplete(APP_VERSION_UPGRADE_KEY);
-          if (isUpgradeComplete) {
-            upgradeComplete.set(true);
-          }
+    /**
+     * Actually fetching messages from TMS.
+     */
+    private List<Message> doFetchMessages(@Nullable String messageId) throws TopicNotFoundException, IOException {
+      List<Message> messages = new ArrayList<>();
+      LOG.trace("Fetching system topic '{}' with messageId '{}'", messageId);
+      MessageFetcher messageFetcher = messagingContext.getMessageFetcher();
+      try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
+                                                                      topic, 100, messageId)) {
+        while (iterator.hasNext() && !stopping) {
+          messages.add(iterator.next());
         }
-        return appMetadataStore;
-      } catch (DatasetManagementException | IOException e) {
-        throw Throwables.propagate(e);
       }
+      return messages;
     }
 
     /**
-     * Loads the message id from storage. Note that this method is already executed inside a transaction.
+     * Returns the publish time encoded in the given message id.
      *
-     * @param context the dataset context
-     * @return the message id, or null if no message id was previously stored
+     * @param messageId the message id to decode
+     * @return the publish time or {@code 0} if the message id is {@code null}.
      */
-    public abstract String loadMessageId(DatasetContext context);
+    private long getMessagePublishTime(@Nullable String messageId) {
+      if (messageId == null) {
+        return 0L;
+      }
 
-    /**
-     * Persists the message id to storage. Note that this method is already executed inside a transaction.
-     *
-     * @param context the dataset context
-     * @param lastFetchedMessageId the message id to persist
-     */
-    public abstract void updateMessageId(DatasetContext context, String lastFetchedMessageId);
+      return new MessageId(Bytes.fromHexString(messageId)).getPublishTimestamp();
```

Comment:
ah never mind, I reviewed commit by commit and did not see this was already done.